src/gf_ring.h

```diff
@@ -490,30 +490,55 @@ T RingModN<T>::weak_rand(void)
     return arith::weak_rand<T>(this->card());
 }
 
-/*
- * Check if a number is primitive root or not, i.e. check its
- *  order y = q - 1, i.e. x^i != 1 for every i < q-1
- *
- * Note, order of a number must be a divisor of (q-1)
- *
- * Algorithm to checking whether x is primitive root or not
- *  Methodology: checking its order is not a proper divisor of (q-1)
- * 1. Prime factorization (q - 1) = p1^r1 * p2^r2 * ... pm^rm
- * 2. Get "necessary" proper divisors of (q - 1):
- *      D = {(q-1)/p1, (q-1)/p2, .., (q-1)/pm }
- * 3. x is a primitive root if for every divisor d of D:
- *                          x^d != 1
+/** Check if a number is primitive root or not.
+ *
+ * A number `x` is a primitive root if its order \f$y = q - 1\f$, i.e.
+ * \f$x^{i}\neq 1\qquad (i=1, 2, 3, \dots , q-1)\f$
+ *
+ * Note, order of a number must be a divisor of q-1.
+ *
+ * Algorithm to checking whether `x` is primitive root or not.
+ *
+ * Methodology: checking its order is not a proper divisor of (q-1).
+ *
+ * 1. Prime factorization \f$q - 1 = p_1^{r_1} p_2^{r_2} \dots p_m^{r_m}\f$
+ * 2. Get "necessary" proper divisors of `q - 1`:
+ *    \f$ D =
+ *    \left\{
+ *      \frac{q-1}{p_1}, \frac{q-1}{p_2}, \dots, \frac{q-1}{p_m}
+ *    \right\}
+ *    \f$
+ * 3. `x` is a primitive root if for every divisor `d` of `D`: \f$x^d \neq 1\f$
+ *
  * Proof:
- *  Suppose that x satisfies the algorithm but its order is a proper divisor y
- *  of (q-1). This order can be expressed as (q-1)/y where
- *              y = (p1^s1 * p2^s2 * ... * pm^sm)
- *  and, x^((q-1)/y) = 1
- *  Withouth loss of generality, we suppose s1 >= 1. We have
- *      x^((q-1)/y) = 1 => (x^((q-1)/y))^(y/p1) = 1 <=> x^((q-1)/p1) = 1.
- *  Contradict to Step 3 of the algorith.
+ *
+ *  Suppose that `x` satisfies the algorithm but its order is a proper divisor
+ *  `y` of `q-1`.
+ *
+ *  This order can be expressed as \f$\frac{q-1}{y}\f$ where:
+ *
+ *  \f[
+ *    y = p_1^{r_1} p_2^{r_2} \dots p_m^{r_m}
```

Comment:
It should be `y = p_1^{s_1} p_2^{s_2} \dots p_m^{s_m}`
with an implicit note: `s_i <= r_i`

Reply:
Indeed, bad copy paste, I'll fix it!